WT-14605 Add statistics for page discard

bzm-jas opened the pull request:
Add page discard statistics to track how many pages are discarded by the disagg block manager. We're only going forward with number of pages discarded in this PR, see details in comments of WT-14605.

bzm-jas committed:
WT-14605 Add statistics for page discard

etienneptl commented on dist/stat_data.py:
```diff
@@ -1199,6 +1199,7 @@ def __init__(self, name, desc, flags=''):
     BlockDisaggStat('disagg_block_hs_get', 'Disaggregated block manager get from the shared history store in SLS'),
     BlockDisaggStat('disagg_block_hs_put', 'Disaggregated block manager put to the shared history store in SLS'),
     BlockDisaggStat('disagg_block_put', 'Disaggregated block manager put '),
+    BlockDisaggStat('disagg_block_pages_discarded', 'Pages discarded by disaggregated block manager'),
```
Please sort them alphabetically
```suggestion
    BlockDisaggStat('disagg_block_pages_discarded', 'Pages discarded by disaggregated block manager'),    BlockDisaggStat('disagg_block_put', 'Disaggregated block manager put '),

```

bzm-jas replied:
Thanks for picking it up, changed in [e09ba12](https://github.com/wiredtiger/wiredtiger/pull/12032/commits/e09ba12391d50aad687479a49b22739a1d2b5960)

etienneptl reviewed (approved):
LGTM % one np

bzm-jas committed:
alphabetical order

bzm-jas committed:
resolve merge conflicts

bzm-jas committed:
re-generate stats

bzm-jas commented:
@pmacko86 Hi Peter, I’m seeing a failure for the unit-test-hook-disagg-leader test on this PR. Could you help double-check if this is a known issue? Thanks!

pmacko86 commented:
> Hi Peter, I’m seeing a failure for the unit-test-hook-disagg-leader test on this PR. Could you help double-check if this is a known issue? Thanks!

Yes, this is a known issue. It sometimes succeeds, but usually fails...

pmacko86 reviewed (approved):
LGTM. Thanks!

bzm-jas committed:
Merge branch 'wiredtiger-disagg-integration' into wt-14605-Add-statistics-for-page-discard

mariammojid commented on src/block_disagg/block_disagg_write.c:
```diff
@@ -303,6 +303,8 @@ __wti_block_disagg_page_discard(
+    WT_STAT_CONN_INCR(session, disagg_block_pages_discarded);
+
     return (
       plhandle->plh_discard(plhandle, &session->iface, page_id, checkpoint_id, &discard_args));
```
shouldn't we increment the `disagg_block_pages_discarded` stat only if `plh_discard` returns success 0? If this stat is intended to count the number of page discard _calls_ then please disregard this comment

etienneptl replied:
Good point, I have suggested edits to reflect that we only capture the number of calls for now.

mariammojid reviewed (approved):
LGTM with a small note

etienneptl commented on dist/stat_data.py:
```diff
@@ -1221,6 +1221,7 @@ def __init__(self, name, desc, flags=''):
     BlockDisaggStat('disagg_block_hs_byte_write', 'Bytes written to the shared history store in SLS', 'size'),
     BlockDisaggStat('disagg_block_hs_get', 'Disaggregated block manager get from the shared history store in SLS'),
     BlockDisaggStat('disagg_block_hs_put', 'Disaggregated block manager put to the shared history store in SLS'),
+    BlockDisaggStat('disagg_block_pages_discarded', 'Pages discarded by disaggregated block manager'),
```
```suggestion
    BlockDisaggStat('disagg_block_page_discard', 'Disaggregated block manager page discard'),
```

etienneptl commented on src/block_disagg/block_disagg_write.c:
```diff
@@ -303,6 +303,8 @@ __wti_block_disagg_page_discard(
     discard_args.backlink_checkpoint_id = checkpoint_id;
     discard_args.backlink_lsn = lsn;
 
+    WT_STAT_CONN_INCR(session, disagg_block_pages_discarded);
```
```suggestion
    WT_STAT_CONN_INCR(session, disagg_block_page_discard);
```

bzm-jas committed:
rename page disagg calls

bzm-jas commented:
@mariammojid @etienneptl Thanks for the suggestions! I've renamed the stat and added clarification that it tracks disagg page discard calls in [f238e26](https://github.com/wiredtiger/wiredtiger/pull/12032/commits/f238e263f50b66d3d42d8d00ce7eb559a6ff66e2)
